### PR TITLE
add pre-commit hook for black

### PR DESCRIPTION
**Pull request by tylerflex**
If you make a commit and black can make formatting changes, the changes are made and the commit isn't registered. Should avoid any commits where black is not applied.

```
pip install pre-commit
pre-commit install
```
I also needed
```
pre-commit run --all-files
```
before it started working.

[see this guide](https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/) for more info.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+repos:
+-   repo: https://github.com/psf/black
+    rev: 22.12.0
+    hooks:
+    - id: black
+      language_version: python3.10
```
Comment (shashwat-sh) [inline .pre-commit-config.yaml:6]: The hard-coded python version caused an error for me on Python 3.8. Worked after I manually changed this to 3.8

**Review by shashwat-sh — COMMENTED** [review]
This will be super useful!

Aside from the python version comment below, only one other observation:
when I tested this with a poorly-formatted line, it worked as expected, but I am concerned it might fool the developer into thinking that the commit went through, because the last thing shown on the terminal is effectively a success message (`"All done! 1 file reformatted."`). Is it possible to print a loud (all caps?) message saying "YOUR CHANGES WERE NOT COMMITTED" or similar?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (shashwat-sh) [reply .pre-commit-config.yaml:6]: Not sure if it can be detected somehow? Or maybe we can get away without specifying the language version?

**Comment by momchil-flex** [comment]
Hmm I thought that this hook would actually `git add` the blacked files. Can't we do that instead of erroring?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (tylerflex) [reply .pre-commit-config.yaml]: good catch, I just removed the language_version entirely and still seems to work. could you try again?

**Comment by tylerflex** [comment]
> Hmm I thought that this hook would actually git add the blacked files and then do the commit. Can't we do that instead of erroring?
@momchil-flex I see what you mean, and I'm not sure. I guess this might be considered unsafe? we might want to just live with it. @joamatab how do you handle this? ie if a pre-commit modifies files?

**Comment by tylerflex** [comment]
@joamatab if I'm understanding you correctly, I already did remove all the checks except for black in this PR. I think momchil is wondering why the black pre-commit hook doesn't `git add` its changes and apply the commit again, which would be a bit more convenient because now a user needs to `git add .` and `git commit` after the pre-commit fails. After reading a little, it seems the main reason is because pre-commit doesn't want to be too aggressive and would rather leave this up to you, but there are some ways to go about it. Just not sure if you use any of them or have suggestions. Personally, I'm fine just using this as a reminder to black code before committing but if there's a better solution, we can include it.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (shashwat-sh) [reply .pre-commit-config.yaml]: Hmm, it no longer does anything, just commits. When I add `language_version` back, then it works as before.

Comment (tylerflex) [reply .pre-commit-config.yaml]: weird. it worked for me. I think `default_language_version` is optional, according to [the docs](https://pre-commit.com/#top_level-default_language_version).  I just tested again with a clean install. maybe you need to do `pre-commit install` again?

Also, I added a custom message to the tool telling the user that if the pre-commit fails, they'll need to add and commit again. I don't think there was a way to trigger this on failure, see [this discussion](https://github.com/pre-commit/pre-commit/issues/1565) for more details.

**Comment by tylerflex** [comment]
@momchil-flex  / @shashwat-sh , if you're ok with this behavior (needing to re-commit with the added warning message), we can merge this. Let me know.

**Comment by shashwat-sh** [comment]
> @momchil-flex / @shashwat-sh , if you're ok with this behavior (needing to re-commit with the added warning message), we can merge this. Let me know.

Looks good to me.

**Comment by momchil-flex** [comment]
seems good to me too!

**Pull request merged**